### PR TITLE
Restart Ironic when certificate is updated

### PR DESCRIPTION
**Pull request by namnx228**
When the TLS certificate is updated, Ironic will reload (or restart) to get the new certificate.

### CODE REVIEW

File: ironic-scripts/runironic-conductor [hunk 1]
```diff
@@ -13,4 +13,12 @@ until ironic-dbsync --config-file /etc/ironic/ironic.conf upgrade; do
   echo "WARNING: ironic-dbsync failed, retrying"
   sleep 1
 done
+
+if [ "$IRONIC_TLS_SETUP" == "true" ]; then
+    inotifywait -m -e delete_self "${IRONIC_CERT_FILE}" | while read file event; do
+     kill $(pgrep ironic)
+     /usr/bin/ironic-conductor &
```
Comment (dtantsur) [inline ironic-scripts/runironic-conductor:20]: No starting anything in the background please. The previous line will likely cause the container to stop anyway.

**Review by dtantsur — CHANGES_REQUESTED** [review]
I don't think this approach will fly, you probably need whatever party orchestrates the containers to restart them.

**Comment by namnx228** [comment]
> I don't think this approach will fly, you probably need whatever party orchestrates the containers to restart them.

Yes, I agree that this approach is not an optimal way. However, at the moment, we are needing this feature soon, so a straightforward solution is good to go, IMO. But I think we will need to design a more K8s native solution in the future.

**Comment by maelk** [comment]
Considering that cert-manager can rotate the certificates at any time, without notifications (it just updates the secret), I don't see a better solution than Ironic being able to detect that the certificate changed and reload it, or since it does not support that for now, having a helper script that will cause a restart of Ironic.

**Comment by s3rj1k** [comment]
inotify would not work on NFS for example

File: prepare-image.sh [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 set -euxo pipefail
 
-dnf install -y python3 python3-requests
+dnf install -y python3 python3-requests epel-release
```
Comment (dtantsur) [inline prepare-image.sh:5]: Please make enabling epel conditional on RESTART_CONTAINER_CERTIFICATE_UPDATED, not everyone wants to have it (and it may even conflict with tripleo repos).

File: ironic-scripts/configure-ironic.sh [hunk 3]
```diff
@@ -8,6 +8,7 @@ export IRONIC_INSECURE=${IRONIC_INSECURE:-false}
 export IRONIC_INSPECTOR_CERT_FILE=/certs/ironic-inspector/tls.crt
 export IRONIC_INSPECTOR_CACERT_FILE=/certs/ca/ironic-inspector/tls.crt
 export IRONIC_INSPECTOR_INSECURE=${IRONIC_INSPECTOR_INSECURE:-$IRONIC_INSECURE}
+export RESTART_CONTAINER_CERTIFICATE_UPDATED=${RESTART_CONTAINER_CERTIFICATE_UPDATED:-"true"}
```
Comment (dtantsur) [inline ironic-scripts/configure-ironic.sh:11]: Should be false, since this is not the traditional way of handling configuration.

**Review by dtantsur — CHANGES_REQUESTED** [review]

File: ironic-scripts/runironic-api [hunk 4]
```diff
@@ -19,5 +19,12 @@ done
 
 python3 -c 'import os; import sys; import jinja2; sys.stdout.write(jinja2.Template(sys.stdin.read()).render(env=os.environ))' < /etc/httpd-ironic-api.conf.j2 > /etc/httpd/conf.d/ironic.conf
 sed -i "/Listen 80/c\#Listen 80" /etc/httpd/conf/httpd.conf
+
+if [ "$IRONIC_TLS_SETUP" == "true"  && "${RESTART_CONTAINER_CERTIFICATE_UPDATED}" == "true" ]; then
+    inotifywait -m -e delete_self "${IRONIC_CERT_FILE}" | while read file event; do
+      kill $(pgrep httpd)
+    done &
```
Comment (zaneb) [inline ironic-scripts/runironic-api:26]: How does this interact with the fact that we are calling `exec` below, so the subshell spawned here becomes a child of the httpd process?

File: ironic-scripts/runironic-api [hunk 5]
```diff
@@ -19,5 +19,12 @@ done
 
 python3 -c 'import os; import sys; import jinja2; sys.stdout.write(jinja2.Template(sys.stdin.read()).render(env=os.environ))' < /etc/httpd-ironic-api.conf.j2 > /etc/httpd/conf.d/ironic.conf
 sed -i "/Listen 80/c\#Listen 80" /etc/httpd/conf/httpd.conf
+
+if [ "$IRONIC_TLS_SETUP" == "true"  && "${RESTART_CONTAINER_CERTIFICATE_UPDATED}" == "true" ]; then
+    inotifywait -m -e delete_self "${IRONIC_CERT_FILE}" | while read file event; do
+      kill $(pgrep httpd)
```
Comment (zaneb) [inline ironic-scripts/runironic-api:25]: This should use `kill -HUP` (for a graceful restart), and I suspect the same is true of the others.

File: ironic-scripts/configure-ironic.sh [hunk 6]
```diff
@@ -8,6 +8,7 @@ export IRONIC_INSECURE=${IRONIC_INSECURE:-false}
 export IRONIC_INSPECTOR_CERT_FILE=/certs/ironic-inspector/tls.crt
 export IRONIC_INSPECTOR_CACERT_FILE=/certs/ca/ironic-inspector/tls.crt
 export IRONIC_INSPECTOR_INSECURE=${IRONIC_INSPECTOR_INSECURE:-$IRONIC_INSECURE}
+export RESTART_CONTAINER_CERTIFICATE_UPDATED=${RESTART_CONTAINER_CERTIFICATE_UPDATED:-"false"}
```
Comment (zaneb) [inline ironic-scripts/configure-ironic.sh:11]: It seems to me that rather than have an environment variable we could just check whether inotifywait exists, and assume that if the functionality was built into the container that we want to use it.

File: ironic-scripts/runmariadb [hunk 7]
```diff
@@ -18,6 +19,13 @@ fi
 
 ln -sf /proc/self/fd/1 /var/log/mariadb/mariadb.log
 
+# Restart mysqld when the certificate is updated
+if [ -f "$MARIADB_CERT_FILE" && "${RESTART_CONTAINER_CERTIFICATE_UPDATED}" == "true" ]; then
+    inotifywait -m -e delete_self "${MARIADB_CERT_FILE}" | while read file event; do
+       kill $(pgrep mysqld)
```
Comment (zaneb) [inline ironic-scripts/runmariadb:25]: `mysqld_safe` automatically restarts if mysqld exits, so this is consistent with the commit message but inconsistent with the implementation of the other services (where we exit the container).

File: prepare-image.sh (hunk 2 above)
Comment (namnx228) [reply prepare-image.sh]: @dtantsur The environment variable RESTART_CONTAINER_CERTIFICATE_UPDATED can only be set when we run the container, but the package needs to be installed when the image is built. Do you have any ideas how should we handle it?

Comment (dtantsur) [reply prepare-image.sh]: You can provide options in Dockerfiles as well

Comment (namnx228) [reply prepare-image.sh]: When we use the image, for example in BMO, we only pull the images from quay.io. We don't rebuild the image using the Dockerfile. In this case, we cannot pick the option that is provided in Dockerfiles.

File: ironic-scripts/runironic-api (hunk 4 above)
Comment (namnx228) [reply ironic-scripts/runironic-api:26]: Yes, I have checked. The subshell becomes a child of the httpd process

File: ironic-scripts/runironic-api (hunk 5 above)
Comment (namnx228) [reply ironic-scripts/runironic-api:25]: I tried to send SIGHUP signal to the ironic-conductor process and I can see that it reloads the config, but it does not re-read the TLS certificate file. So the only way is to restart the container.

File: ironic-scripts/configure-ironic.sh (hunk 6 above)
Comment (namnx228) [reply ironic-scripts/configure-ironic.sh:11]: I think the inotify-tools will always be baked into the image, but it is the user's choice to use it or not. That is why we need this environment variable.

File: ironic-scripts/runironic-api (hunk 5 above)
Comment (s3rj1k) [reply ironic-scripts/runironic-api:25]: just a thought, how about ditching apache in favor of nginx? 
(nginx can dynamically load SSL certs)

Comment (namnx228) [reply ironic-scripts/runironic-api:25]: Btw, thanks for the suggestion to use `kill -WINCH`. It is helpful.

Comment (namnx228) [reply ironic-scripts/runironic-api]: @s3rj1k Apache server can also do the same, actually. The reason why I restart the httpd is to make it consistent with the other containers

File: ironic-scripts/runmariadb (hunk 7 above)
Comment (namnx228) [reply ironic-scripts/runmariadb]: Thanks! Fixed!

File: prepare-image.sh (hunk 2 above)
Comment (dtantsur) [reply prepare-image.sh]: Well, you have to rebuild it or publish an alternative image. Or you can only install inotify from EPEL without pulling in anything else.
```
09:45 <dtantsur> hi folks! I remember long ago it was not recommended to use EPEL with OpenStack repositories. What's the situation now?
10:26 <apevec> dtantsur, the same
```

File: prepare-image.sh [hunk 8]
```diff
@@ -2,9 +2,11 @@
 
 set -euxo pipefail
 
-dnf install -y python3 python3-requests
-curl https://raw.githubusercontent.com/openstack/tripleo-repos/master/tripleo_repos/main.py | python3 - -b master current-tripleo
+dnf install -y python3 python3-requests epel-release 
+
+curl https://raw.githubusercontent.com/openstack/tripleo-repos/master/tripleo_repos/main.py | python3 - -b master current-tripleo --no-stream
```
Comment (dtantsur) [inline prepare-image.sh:7]: The `--no-stream` change is unrelated and should be posted separately.

File: prepare-image.sh [hunk 9]
```diff
@@ -2,9 +2,11 @@
 
 set -euxo pipefail
 
-dnf install -y python3 python3-requests
-curl https://raw.githubusercontent.com/openstack/tripleo-repos/master/tripleo_repos/main.py | python3 - -b master current-tripleo
+dnf install -y python3 python3-requests epel-release
```
[the remark this answers is not in the document]
Comment (namnx228) [reply prepare-image.sh:5]: Ok, I have updated that only the inotify-tools package is installed from EPEL

**Comment by namnx228** [comment]
This PR needs to wait for https://github.com/metal3-io/ironic-image/pull/259 to go in before it can pass the Travis CI test

File: prepare-image.sh [hunk 10]
```diff
@@ -11,6 +11,7 @@ if [[ ! -z ${EXTRA_PKGS_LIST:-} ]]; then
         xargs -rtd'\n' dnf --setopt=install_weak_deps=False install -y < /tmp/${EXTRA_PKGS_LIST}
     fi
 fi
+dnf install -y https://download-ib01.fedoraproject.org/pub/epel/8/Everything/x86_64/Packages/i/inotify-tools-3.14-19.el8.x86_64.rpm
```
Comment (dtantsur) [inline prepare-image.sh:14]: This will break once this version is deleted. What you need is to install epel-release but disable the added repo via dnf (`dnf config-manager --set-disabled epel` or something like that). Then you can use `--enablerepo=epel` when installing inotify-tools specifically.

**Review by dtantsur — APPROVED** [review]

**Review by maelk — COMMENTED** [review]
/lgtm
